cfstore/cfstore.go

```diff
@@ -0,0 +1,204 @@
+package cfstore
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"code.cloudfoundry.org/lager"
+	cfclient "github.com/cloudfoundry-community/go-cfclient"
+)
+
+const (
+	DefaultInitTimeout = 5 * time.Minute
+)
+
+type Config struct {
+	// CFClient config
+	ClientConfig *cfclient.Config
+	// Client for communicating with cf
+	Client CFDataClient
+	// Database connection
+	DB *sql.DB
+	// Logger overrides the default logger
+	Logger lager.Logger
+	// Collection delay
+	Schedule time.Duration
+}
+
+type Store struct {
+	client CFDataClient
+	db     *sql.DB
+	logger lager.Logger
+}
+
+func (s *Store) Init() error {
+	s.logger.Info("initializing")
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultInitTimeout)
+	defer cancel()
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	if err := s.collectServices(tx); err != nil {
+		return err
+	}
+	if err := s.collectServicePlans(tx); err != nil {
+		return err
+	}
+	s.logger.Info("initialized")
+	return tx.Commit()
+}
+
+func (s *Store) CollectServicePlans() error {
+	ctx, cancel := context.WithTimeout(context.Background(), DefaultInitTimeout)
+	defer cancel()
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	if err := s.collectServicePlans(tx); err != nil {
+		return err
+	}
+	return tx.Commit()
+}
+
+func (s *Store) collectServicePlans(tx *sql.Tx) error {
+	plans, err := s.client.ListServicePlans()
+	if err != nil {
+		return err
+	}
+	for _, plan := range plans {
+		validFrom := plan.UpdatedAt
+		var planCount int
+		err := tx.QueryRow(`
+			select count(*)
+			from service_plans
+			where guid = $1
+		`, plan.Guid).Scan(&planCount)
+		if err != nil {
+			return err
+		}
+		if planCount == 0 {
+			validFrom = plan.CreatedAt
+		}
+
+		var serviceValidFrom *time.Time
+		err = tx.QueryRow(`
+			select valid_from
+			from services
+			where guid = $1
+			order by valid_from desc
+			limit 1
+		`, plan.ServiceGuid).Scan(&serviceValidFrom)
+		if err == sql.ErrNoRows {
+			panic("meh dunno what to do")
```

Comment:
Do we want to panic here? Can we not:

```
if err != nil {
  return err
}
```

for the error to be handled properly later on and not kill the application?

Reply:
yep def shouldn't panic that shouldn't have been left in